lib/sources.nix

```diff
@@ -26,12 +26,28 @@ rec {
     (type == "symlink" && lib.hasPrefix "result" baseName)
   );
 
-  cleanSource = builtins.filterSource cleanSourceFilter;
+  cleanSource = filterSource cleanSourceFilter;
+
+  # Like `builtins.filterSource`, except it will compose with itself,
+  # allowing you to chain multiple calls together without any
+  # intermediate copies being put in the nix store.
+  filterSource = f: src:
+    let isPath = builtins.typeOf src == "path";
```

Comment:
Seems like it would be better to switch on `src ? __is-lib-filterSource` than `isPath`. E.g. hydra will pass in paths as `{ outPath = the-actual-path; etc. }` so this will spuriously fail here.

Reply:
Wouldn't builtins.filterSource fail on such inputs as well? AFAIK, this will succeed on anything builtins.filterSource succeeds on

Reply:
`filterSource` calls `coerceToPath`, which respects `outPath` etc.

Reply:
Just pushed a change. Is that better?